enhancements/dns/externaldns-operator-aws-assume-role.md

```diff
@@ -0,0 +1,287 @@
+4. **shared vpc admin** creates an IAM role in Account A with a Trust Policy granting Account B IAM User to assume it,
+   and attach the custom IAM policy created in step 3.
+5. **cluster admin** installs the cluster in Account B.
+6. **cluster admin** installs ExternalDNS Operator in the new cluster.
+7. **cluster admin** specifies the role from step 4 in a new ExternalDNS CR object.
+
+### API Extensions
+
+We extend the`ExternalDNS` API object by adding the struct `AssumeRole` to the `ExternalDNSAWSProviderOptions` struct
+(`spec.provider.aws`). Refer to [`externaldns_types.go`](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go)
+for the existing API structure of the `ExternalDNS` object.
+
+The ExternalDNS Operator's validation webhook will use the existing [`IsARN`](https://pkg.go.dev/github.com/aws/aws-sdk-go/service/s3/internal/arn#IsARN)
+function to validate the `ARN` field in the `AssumeRole` struct (`spec.provider.aws.assumeRole.arn`). The approach
+mentioned above differs from the design documented in the [API: DNS](/enhancements/installer/aws-cross-account-dns-zone.md#API-DNS)
+section of the "AWS Shared VPC with Cross-account DNS Zones" enhancement, which employs CRD validation utilizing a regular
+expression. Using the existing `IsARN` function provides more flexibility and reliability since it is maintained by AWS
+themselves, which helps to mitigate potential issues with edge cases.
+
+Furthermore, we introduce a default value for `spec.provider.aws.credentials.name`, providing OpenShift users the
+ability to utilize the `AssumeRole` field without the necessity of specifying a `Credentials` field. For additional
+information regarding this update, please refer to [The Required Credentials Issue](#the-required-credentials-issue).
+
+```go
+type ExternalDNSAWSProviderOptions struct {
+    // Credentials is a reference to a secret containing
+    // the following keys (with corresponding values):
+    //
+    // * aws_access_key_id
+    // * aws_secret_access_key
+    //
+    // See
+    // https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md
+    // for more information.
+    //
+    // +kubebuilder:validation:Required
+    // +kubebuilder:default:={"name":""}
+    // +required
+    Credentials SecretReference `json:"credentials"`
+
+    // assumeRole is a reference to the IAM role that
+    // ExternalDNS will be assuming in order to perform
+    // any DNS updates.
+    //
+    // +kubebuilder:validation:Optional
+    // +optional
+    AssumeRole *ExternalDNSAWSAssumeRoleOptions `json:"assumeRole,omitempty"`
+}
+
+type ExternalDNSAWSAssumeRoleOptions struct {
+    // arn is an IAM role ARN that the ExternalDNS
+    // operator will assume when making DNS updates.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    ARN string `json:"arn,omitempty"`
```

Comment:
It occurs to me that we may need to plan for multiple ARNs.  Is there any reason to leave this as a single-value field?

Reply:
Fair question. An ARN represents another IAM Role in another AWS account. If there were multiple ARNs, it would imply that External DNS should assume multiple roles. I can only think of 1 reason to assume multiple roles: to create the same DNS record in different accounts.

The correct answer to satisfy the motivation of creating the same DNS record in multiple AWS accounts with multiple RoleARNs would be: create multiple `ExternalDNS` objects so you have multiple instances of External DNS running to service multiple AWS accounts. I don't know of a use case this would benefit, but my point being, we can still support multiple AWS Accounts (i.e. multiple RoleARN) through multiple instances.

Furthermore, the design of External DNS's argument [`--aws-assume-role`](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/pkg/apis/externaldns/types.go#L440) doesn't support multiple ARNs. And on top of that, OpenShift is designed to work with [one RoleARN](https://github.com/openshift/enhancements/blob/5cd4edc75ad6144a0b751cb9700d421b5cce217a/enhancements/installer/aws-cross-account-dns-zone.md?plain=1#L138) provided at install time.

I think it's safe to say we don't need to plan for multiple ARNs. Let me know you have information that would make this invalid.